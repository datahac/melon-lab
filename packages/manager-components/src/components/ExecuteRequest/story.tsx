import { storiesOf } from '@storybook/react';
import React from 'react';
import ExecuteRequest from './index';

const data = {};

storiesOf('Components|Execute Request', module).add('Default', () => {
  return <ExecuteRequest {...data} />;
});
