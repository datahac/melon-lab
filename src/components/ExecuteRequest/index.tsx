import React from 'react';
import ModalTransaction from '+/components/ModalTransaction';
import { withRouter } from 'next/router';
import gql from 'graphql-tag';

const estimateExecuteRequestMutation = gql`
  mutation EstimateExecuteRequest($fundAddress: String!) {
    estimate: estimateExecuteRequest(fundAddress: $fundAddress) @account {
      data
      from
      gas
      gasPrice
      to
      value
    }
  }
`;

const executeExecuteRequestMutation = gql`
  mutation ExecuteExecuteRequest(
    $data: String!
    $from: String!
    $gas: String!
    $gasPrice: String!
    $to: String!
    $value: String!
    $fundAddress: String!
  ) {
    execute: executeExecuteRequest(
      fundAddress: $fundAddress
      unsigned: {
        data: $data
        from: $from
        gas: $gas
        gasPrice: $gasPrice
        to: $to
        value: $value
      }
    ) @sign @account
  }
`;

export default withRouter(props => (
  <ModalTransaction
    text="The following method on the Melon Smart Contracts will be executed: executeRequest"
    open={props.step === 2}
    estimate={{
      mutation: estimateExecuteRequestMutation,
      variables: () => ({
        fundAddress: props.fundAddress,
      }),
    }}
    execute={{
      mutation: executeExecuteRequestMutation,
      variables: (_, transaction) => ({
        ...transaction,
        fundAddress: props.fundAddress,
      }),
      update: () => {
        props.setStep(3);
      },
      onCompleted: () => {
        props.router.push({
          pathname: '/manage',
          query: {
            address: props.fundAddress,
          },
        });
      },
    }}
    handleCancel={() => {
      props.setStep(null);
    }}
  />
));
