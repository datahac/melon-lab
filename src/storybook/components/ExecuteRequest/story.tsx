import { action } from '@storybook/addon-actions';
import { storiesOf } from '@storybook/react';
import React from 'react';
import ExecuteRequest from './index';

const data = {
  onExecute: action('onExecute'),
  requestId: 1,
};

storiesOf('Components|Execute Request', module)
  .add('Default', () => {
    return <ExecuteRequest {...data} readyToExecute={true} />;
  })
  .add('No ready', () => {
    return <ExecuteRequest {...data} />;
  });
