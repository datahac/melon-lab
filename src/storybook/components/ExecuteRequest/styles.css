.execute-request__title {
  margin-top: 0;
}
