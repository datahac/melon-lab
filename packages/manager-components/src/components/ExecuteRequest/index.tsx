import React, { Fragment, StatelessComponent } from 'react';
// import Countdown from 'react-cntdwn';
import Button from '~/blocks/Button';
import Notification from '~/blocks/Notification';

import styles from './styles.css';

export interface ExecuteRequestProps {
  onExecute?: (id) => void;
  readyToExecute?: boolean;
  targetDate;
  requestId?: string;
}

export const ExecuteRequest: StatelessComponent<ExecuteRequestProps> = ({
  onExecute,
  readyToExecute,
  targetDate,
  requestId,
}) => {
  const onClickExecute = () => onExecute && onExecute(requestId);

  return (
    <div className="execute-request">
      <Notification isWarning>
        <style jsx>{styles}</style>
        <h3 className="execute-request__title">Waiting time required</h3>
        <p>
          <a
            href="https://medium.com/melonport-blog/protecting-participants-ee55a752287"
            target="_blank"
            rel="noopener noreferrer"
          >
            To prevent information advantage of any economic agent over another,
            we enforce a waiting period before any subscription or redemption
            can be executed.
          </a>
        </p>

        {readyToExecute ? (
          <Button style="secondary" onClick={onClickExecute}>
            Execute my request!
          </Button>
        ) : (
          <Fragment>
            Remaining waiting time before request execution:
            {/* <Countdown
                targetDate={targetDate}
                startDelay={100}
                interval={1000}
                timeSeparator=":"
                leadingZero
              /> */}
          </Fragment>
        )}
      </Notification>
    </div>
  );
};

export default ExecuteRequest;
