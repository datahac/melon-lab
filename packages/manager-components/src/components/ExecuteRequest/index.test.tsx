import React from 'react';
import ExecuteRequest from './index';

const data = {
  onExecute: () => null,
  readyToExecute: true,
  requestId: 1,
};

describe('ExecuteRequest', () => {
  const defaultElement = <ExecuteRequest {...data} />;

  it('should render correctly', () => {
    const wrapper = shallow(defaultElement);
    expect(wrapper).toMatchSnapshot();
  });
});
