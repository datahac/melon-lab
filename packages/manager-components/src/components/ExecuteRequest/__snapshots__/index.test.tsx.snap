// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExecuteRequest should render correctly 1`] = `
<div
  className="execute-request"
>
  <Notification
    isWarning={true}
  >
    <JSXStyle
      css="__scoped"
      styleId="__scopedHash"
    />
    <h3
      className="jsx-__scopedHash execute-request__title"
    >
      Waiting time required
    </h3>
    <p
      className="jsx-__scopedHash"
    >
      <a
        className="jsx-__scopedHash"
        href="https://medium.com/melonport-blog/protecting-participants-ee55a752287"
        rel="noopener noreferrer"
        target="_blank"
      >
        To prevent information advantage of any economic agent over another, we enforce a waiting period before any subscription or redemption can be executed.
      </a>
    </p>
    <Button
      onClick={[Function]}
      style="secondary"
    >
      Execute my request!
    </Button>
  </Notification>
</div>
`;
